core/src/main/java/jenkins/model/lazy/AbstractLazyLoadRunMap.java

```diff
@@ -336,9 +335,9 @@ public boolean runExists(int number) {
             return null;
         case DESC:
             // TODO again could be made more efficient
-            List<Integer> reversed = new ArrayList<Integer>(numberOnDisk);
-            Collections.reverse(reversed);
-            for (int m : reversed) {
+            ListIterator<Integer> iterator = numberOnDisk.listIterator(numberOnDisk.size());
```

Comment:
@Jimilian I think this is a good start, but better to use the 'find' method which doesn't do unboxing and  will use more efficient binary search to find the index the item would be inserted at (well, technically `-(insertionPoint+1)`)https://github.com/jenkinsci/jenkins/blob/12fd3e47e393fb1942cd5cb29766bcefe62a81f8/core/src/main/java/jenkins/model/lazy/SortedIntList.java#L62.  In fact I think both impls should use that instead - the SortedIntList also has an old  impl of binary search that could be replaced with the stock JVM one.

Reply:
but we do not know the exact number we want to get. do you suggest to find one-by-one?

Reply:
@Jimilian The index returned should be at most one index entry away from the desired value since it's finding the insertion point for that item (it'll either be one slot up or down).   If desired entry is in the list, it'll match exactly.

Reply:
No, it could be many index entries away, since `getByNumber` can fail, so you would still need to iterate.

Also I _think_ you meant `floor` not `find`.

(Yes `SortedIntList.binarySearch` could be replaced now.)

Reply:
Ah, good point about getByNumber -- but I think this will still get you much closer to the target than direct iteration, no?

And yes, you can use `floor`/`ceil` here to further simplify the search-below / search-above methods.

Reply:
@svanoort, we have next use case in Core:
```java
hudson.model.RunMap
=> newestValue()
        return search(Integer.MAX_VALUE, DESC);
=> oldestValue()
        return search(Integer.MIN_VALUE, ASC);

jenkins.model.lazy.AbstractLazyLoadRunMap
=> isEmpty()
        return search(Integer.MAX_VALUE, DESC)==null;

=> newestBuild()
        return search(Integer.MAX_VALUE, DESC);
=> oldestBuild()
        return search(Integer.MIN_VALUE, ASC);

=> subMap(Integer, Integer)
        R start = search(fromKey, DESC);
        R end = search(toKey, ASC);
        i = search(getNumberOf(i)-1,DESC);

AbstractLazyLoadRunMapTest.LazyBuildMixIn
=> getNearestBuild(int)
        return builds.search(n, AbstractLazyLoadRunMap.Direction.ASC);
=> getNearestOldBuild(int)
        return builds.search(n, AbstractLazyLoadRunMap.Direction.DESC);

LazyBuildMixIn.RunMixIn
=> getNextBuild()
                    RunT nb = asRun().getParent().getLazyBuildMixIn()._getRuns().search(asRun().number + 1, AbstractLazyLoadRunMap.Direction.ASC);
=> getPreviousBuild()
                    RunT pb = _parent.getLazyBuildMixIn()._getRuns().search(asRun().number - 1, AbstractLazyLoadRunMap.Direction.DESC);

LazyLoadRunMapEntrySet
=> iterator()
                    next = owner.search(owner.getNumberOf(last)-1, Direction.DESC);
```

In some cases current approach will be faster than binarySearch (`INTEGER.MAX_VALUE` and `INTEGER.MIN_VALUE`). And most of the time during cpu/memory profiling of my production instance I observe `getPreviousBuild`. So, I can add support for all cases (i.e. check the target value and switch to iteration if it's needed), but before I will invest to it I want to know that everyone is ok with this approach (at least current reviewers: @svanoort, @jglick, @oleg-nenashev).